lib/accesslists/collection.go

```diff
@@ -0,0 +1,151 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package accesslists
+
+import (
+	"context"
+	"slices"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types/accesslist"
+)
+
+// Collection represents a complete batch of access from one source (e.g. entraID or okta)
+// That implies that all access lists and members in the batch are expected to be created/updated together
+// because snapshot have a final complete state of the access lists and their members hierarchy.
+// For instance imagine an access list import from EntraID directory where  the access list
+// can't be modified outside EntraID and teleport only fetches the complete state of access lists and members from EntraID
+// This is useful to validate nested hierarchy on the fly and preset the reference updates (Status.MemberOf and Status.OwnerOf)
+// before starting the actual creation/update process in the backend.
+// Where the relation between access list outside the Collection is not possible because collection source are
+// aware only of the access lists and members they contain.
+type Collection struct {
```

Comment:
Nit: "Collection" conflict with the `tools/tctl/common/resources.Collection` interface (we also have an accesslist collection in there). This is not a blocker, but maybe we could find another name.

Reply:
I don’t think this is an issue. The pattern like okta.Service and entraid.Service where the internal Exporter structs share the same name is quite popular and I don't think that  is thi considered as antipatter. . Similarly, for access lists we already work with both proto.AccessList and typeaccesslist.AccessList types without problems.